app/models/concerns/resource_tools/type_dictionary.rb

```diff
@@ -10,19 +10,10 @@ module ResourceTools::TypeDictionary
     attr_accessor :default_description, :preregistration_required
 
     def for_key(key)
-      tries = 0
-      begin
-        if preregistration_required?
-          find_by(key: key)
-        else
-          create_with(description: default_description).find_or_create_by!(key: key)
-        end
-      rescue ActiveRecord::RecordNotUnique => e
-        # If we have two similar events arriving at the same time, we might conceivably run into a
-        # race condition. We retry a few times, then give up and re-raise our exception;
-        tries += 1
-        retry if tries <= RECORD_NOT_UNIQUE_RETRIES
-        raise e
```

Comment:
This is all handled at the message level now. Which is good, as that's where the problem actually manifests. The transaction will helpfully ensure that we don't actually see the exception here.